expression/expression.go

```diff
@@ -309,6 +302,96 @@ func VecEvalBool(ctx sessionctx.Context, exprList CNFExprs, input *chunk.Chunk,
 	return selected, nulls, nil
 }
 
+func toBool(sc *stmtctx.StatementContext, eType types.EvalType, buf *chunk.Column, sel []int, areZeros []int8) error {
+	var err error
+	switch eType {
+	case types.ETInt:
+		i64s := buf.Int64s()
+		for i := range sel {
+			if buf.IsNull(i) {
+				areZeros[i] = -1
+			} else {
+				if i64s[i] == 0 {
+					areZeros[i] = 0
+				} else {
+					areZeros[i] = 1
+				}
+			}
+		}
+	case types.ETReal:
+		f64s := buf.Float64s()
+		for i := range sel {
+			if buf.IsNull(i) {
+				areZeros[i] = -1
+			} else {
+				if types.RoundFloat(f64s[i]) == 0 {
+					areZeros[i] = 0
+				} else {
+					areZeros[i] = 1
+				}
+			}
+		}
+	case types.ETDuration:
+		d64s := buf.GoDurations()
+		for i := range sel {
+			if buf.IsNull(i) {
+				areZeros[i] = -1
+			} else {
+				if d64s[i] == 0 {
+					areZeros[i] = 0
+				} else {
+					areZeros[i] = 1
+				}
+			}
+		}
+	case types.ETDatetime, types.ETTimestamp:
+		t64s := buf.Times()
+		for i := range sel {
+			if buf.IsNull(i) {
+				areZeros[i] = -1
+			} else {
+				if t64s[i].IsZero() {
+					areZeros[i] = 0
+				} else {
+					areZeros[i] = 1
+				}
+			}
+		}
+	case types.ETString:
+		for i := range sel {
+			if buf.IsNull(i) {
+				areZeros[i] = -1
+			} else {
+				iVal, err1 := types.StrToInt(sc, buf.GetString(i))
+				err = err1
+				if iVal == 0 {
+					areZeros[i] = 0
+				} else {
+					areZeros[i] = 1
+				}
+			}
+		}
+	case types.ETDecimal:
+		d64s := buf.Decimals()
+		for i := range sel {
+			if buf.IsNull(i) {
+				areZeros[i] = -1
+			} else {
+				v, err1 := d64s[i].ToFloat64()
+				err = err1
+				if types.RoundFloat(v) == 0 {
+					areZeros[i] = 0
+				} else {
+					areZeros[i] = 1
+				}
+			}
+		}
+	case types.ETJson:
+		return errors.Errorf("cannot convert type json.BinaryJSON to bool")
```

Comment:
It's weird here, please take a look at the example below in mysql8.0.
Maybe it is an issue already existed.

```mysql
mysql> select * from testjson;
+----+-------------+
| id | j           |
+----+-------------+
|  1 | {"test": 1} |
|  2 | {"test": 0} |
+----+-------------+
2 rows in set (0.00 sec)

mysql> select * from testjson where j;
+----+-------------+
| id | j           |
+----+-------------+
|  1 | {"test": 1} |
|  2 | {"test": 0} |
+----+-------------+
2 rows in set (0.00 sec)
```